pkg/arvo/sys/vane/khan.hoon

```diff
@@ -0,0 +1,214 @@
+          $:  %khan                                     ::    from self
+              $>  ?(%arow %avow)  gift                  ::  thread result
+      ==  ==                                            ::
+    +$  khan-state                                      ::
+      [%2 hey=duct]                                     ::  current unix duct
+    --                                                  ::
+=>
+|%
+++  get-beak
+  |=  [=bear now=@da]
+  ?@(bear [our bear %da now] bear)
+++  get-dais
+  |=  [=beak =mark rof=roof]
+  ^-  dais:clay
+  ?~  ret=(rof ~ %cb beak /[mark])
+    ~|(%mark-unknown !!)
+  ?~  u.ret
+    ~|(%mark-invalid !!)
+  ?>  =(%dais p.u.u.ret)
+  !<(dais:clay q.u.u.ret)
+++  make-wire
+  |=  [=beak =mark]
+  ^-  wire
+  :~  %fyrd
+      (scot %p p.beak)
+      q.beak
+      -.r.beak
+      ?-  -.r.beak
+        %da   (scot %da p.r.beak)
+        %tas  p.r.beak
+        %ud   (scot %ud p.r.beak)
+      ==
+      mark
+  ==
+++  read-wire
+  |=  =wire
+  ^-  (pair beak mark)
+  :_  (snag 5 wire)
+  :+  (slav %p (snag 1 wire))
+    (snag 2 wire)
+  =/  ras  (snag 4 wire)
+  ?+  (snag 3 wire)  !!
+    %da   [%da (slav %da ras)]
+    %tas  [%tas ras]
+    %ud   [%ud (slav %ud ras)]
+  ==
+++  start-spider
+  |=  [our=@p =vase]
+  ^-  note
+  [%g %deal [our our] %spider %poke %spider-start vase]
+++  watch-spider
+  |=  [our=@p =path]
+  ^-  note
+  [%g %deal [our our] %spider %watch path]
+--
+=|  khan-state
+=*  state  -
+|=  [now=@da eny=@uvJ rof=roof]
+=*  khan-gate  .
+^?
+|%
+::  +call: handle a +task request
+::
+++  call
+  |=  $:  hen=duct
+          dud=(unit goof)
+          wrapped-task=(hobo task)
+      ==
+  ^-  [(list move) _khan-gate]
+  ::
+  =/  =task  ((harden task) wrapped-task)
+  ?^  dud
+    ~|(%khan-call-dud (mean tang.u.dud))
+  ?+    -.task  [~ khan-gate]
+      %born
+    [~ khan-gate(hey hen)]
+  ::
+      %fard
+    =/  tid=@ta
+      (cat 3 'khan-fyrd--' (scot %uv (sham eny)))
+    =*  fyd     p.task
+    =/  =beak   (get-beak bear.fyd now)
+    =/  args    [~ `tid beak name.fyd args.fyd]
+    :_  khan-gate
+    %+  turn
+      :~  (watch-spider our /thread-result/[tid])
+          (start-spider our !>(args))
+      ==
+    |=(=note ^-(move [hen %pass //g note]))
+  ::
+      %fyrd
+    =*  fyd     p.task
+    =/  =beak   (get-beak bear.fyd now)
+    =/  =wire   (make-wire beak p.args.fyd)
+    =/  =dais:clay
+      (get-dais beak p.q.args.fyd rof)
+    =/  =vase
+      (slam !>(some) (vale.dais q.q.args.fyd))
+    :_  khan-gate
+    [hen %pass wire %k %fard bear.fyd name.fyd vase]~
+  ==
+::  +load: migrate an old state to a new khan version
+::
+++  load
+  |=  $=  old
+      $%  [?(%0 %1) hey=duct *]
+          khan-state
+      ==
+  ^+  khan-gate
+  =/  new=khan-state
+    ?:  ?=(%2 -.old)
+      old
+    [%2 hey.old]
+  khan-gate(state new)
+::  +scry: nothing to see as yet
+::
+++  scry
+  ^-  roon
+  |=  [lyc=gang car=term bem=beam]
+  ^-  (unit (unit cage))
+  ~
```

Comment:
Does `|mass` work without implementing `/whey`?  I don't remember.

Reply:
Maybe? I see:

```
    %khan:
      %dot: B/344
      %typ: KB/92.540
      %sac: KB/8.088
    --KB/100.972
```

Reply:
I made it optional recently.